src/boot/efi/stub.c

```diff
@@ -158,9 +158,13 @@ static bool use_load_options(
         if (secure_boot_enabled() && (have_cmdline || is_confidential_vm()))
                 return false;
 
-        /* We also do a superficial check whether first character of passed command line
-         * is printable character (for compat with some Dell systems which fill in garbage?). */
-        if (loaded_image->LoadOptionsSize < sizeof(char16_t) || ((char16_t *) loaded_image->LoadOptions)[0] <= 0x1F)
+        /* Detect and correct when the Loaded Image Protocol is incorrectly implemented (as in some Dell
+         * firmware) and LoadOptions contains the entire EFI_LOAD_OPTION descriptor, rather than just the
+         * OptionalData field. */
+        efi_apply_loadoptions_quirk((const void **)&(loaded_image->LoadOptions),
+                        &(loaded_image->LoadOptionsSize));
```

Comment:
It looks pretty ugly to get addresses of fields this way. I'm not a C programmer... Suggestions?